In [1]:
import numpy as np
from scipy import stats

### Задача 1

##### Известно, что генеральная совокупность распределена нормально с известным средним квадратическим отклонением 16. Найти доверительный интервал для оценки математического ожидания mu с надёжностью 0.95, если выборочное среднее равно 80, а объём равен 256.

$
Z = \cfrac{\overline{X} - \mu}{\sigma / \sqrt{n}}
$

$
P \left( \overline{X} + t_{a/2} \cdot \cfrac{\sigma}{\sqrt{n}} \leq \mu \leq \overline{X} + t_{1-a/2} \cdot \cfrac{\sigma}{\sqrt{n}} \right) = p
$

In [2]:
sigma = 16
X = 80
n = 256
p = 0.95
mu = '?'

In [3]:
def confidence_interval(X, n, p, sigma):
    '''looking for confidence interval'''
    
    df = n-1
    alpha = 1 - p
    t1 = stats.t.ppf(alpha/2, df)
    t2 = stats.t.ppf(1-alpha/2, df)
    
    return X + t1*(sigma/n**(1/2)), X + t2*(sigma/n**(1/2))

In [4]:
confidence_interval(X, n, p, sigma)

(78.0306894301508, 81.9693105698492)

___Ответ:___ Доверительный интервал для оценки mu с надёжностью 0.95, от 78.03 до 81.97.

### Задача 2

##### Продавец утверждает, что средний вес пачки печенья составляет 200 г. Из партии извлечена выборка из 10 пачек. Вес каждой пачки составляет:

202, 203, 199, 197, 195, 201, 200, 204, 194, 190

##### Известно, что их веса распределены нормально.
    1. Верно ли утверждение продавца, если учитывать, что уровень значимости равен 1%?
    2. Найдите P-значение для данного теста.

##### 1. Верно ли утверждение продавца, если учитывать, что уровень значимости равен 1%?

In [5]:
samples = np.array([202, 203, 199, 197, 195, 201, 200, 204, 194, 190])

mu = 200 #gr
n = 10
alpha = 0.01

In [6]:
t1 = stats.t.ppf(alpha / 2, df=n-1)
t2 = stats.t.ppf(1 - alpha /2, df=n-1)

t1, t2

(-3.24983554401537, 3.2498355440153697)

$
t = \dfrac{\overline{X} - \mu}{\sigma_X / \sqrt{n}}
$

In [7]:
def statistic(samples: np.ndarray, mu):
    '''Z or t'''
    
    sigma = samples.std(ddof=1)
    n = samples.shape[0]
    
    return (samples.mean() - mu)/sigma/n**(1/2)

In [8]:
S = statistic(samples, mu)
S

-0.10651074037450894

In [9]:
t1<=S<=t2

True

___Ответ:___ 1. S не попало в критические область - утверждение продавца верно

##### 2. Найдите P-значение для данного теста.

In [10]:
p_left = stats.t.cdf(S, df=n - 1)
p_right = 1 - stats.t.cdf(S, df=n - 1)

pvalue = 2 * min(p_left, p_right)

pvalue

0.9175134909296933

### Задача 3

##### Дана выборка диаметров подшипников из примера 1 с занятия 5:

samples = [0.6603, 0.9466, 0.5968, 1.3792, 1.5481, 0.7515, 1.0681, 1.1134,
1.2088, 1.701 , 1.0282, 1.3579, 1.0191, 1.1784, 1.1168, 1.1372,
0.7273, 1.3958, 0.8665, 1.5112, 1.161 , 1.0232, 1.0865, 1.02 ]

##### Предполагая, что диаметры подшипников распределены нормально, проверьте гипотезу о том, что дисперсия случайной величины равна 0.0625 при уровне значимости alpha = 0.05. Что для этого нужно знать:
    1. Альтернативная гипотеза двухсторонняя.
    2. Статистика для теста: H = (n - 1) * sample_variance / variance, где n - число элементов в выборке, sample_variance - несмещённая оценка дисперсии, variance - утверждаемая нулевой гипотезой дисперсия.
    3. Эта статистика в предположении верности нулевой гипотезы имеет распределение хи-квадрат с параметром df = n - 1. Её квантили можно найти с помощью функции scipy.stats.chi2.ppf.

In [11]:
samples = np.array([0.6603, 0.9466, 0.5968, 1.3792, 1.5481, 0.7515, 1.0681, 1.1134, 1.2088, 1.701 , 1.0282, 1.3579, 
           1.0191, 1.1784, 1.1168, 1.1372, 0.7273, 1.3958, 0.8665, 1.5112, 1.161 , 1.0232, 1.0865, 1.02 ])
Dx0 = 0.0625
alpha = 0.05

    1. Альтернативная гипотеза двухсторонняя: 
$H_0 = 1mm\\ 
H_1 < 1mm < H_1$

In [30]:
def H(samples: np.ndarray, Dx0):
    '''statistic'''
    
    n = samples.shape[0]
    sample_variance = samples.var(ddof=1)
    variance = Dx0
    
    return (n - 1) * sample_variance / variance

In [37]:
def confidence_interval2(samples: np.ndarray, alpha):
    
    n = samples.shape[0]
    sample_variance = samples.var(ddof=1)
    
    t1 = stats.chi2.ppf(alpha / 2, df=n-1)
    t2 = stats.chi2.ppf(1 - alpha / 2, df=n-1)
    
    return (n-1)*sample_variance / t1, (n-1)*sample_variance / t2

In [33]:
S = H(samples=samples, Dx0=Dx0)
S

28.720541753333343

In [27]:
t1 = stats.chi2.ppf(alpha / 2, df=n-1)
t2 = stats.chi2.ppf(1 - alpha / 2, df=n-1)
t1, t2

(2.7003894999803584, 19.02276779864163)

In [34]:
t1 <= S <= t2

False

In [40]:
ci = confidence_interval2(samples=samples, alpha=alpha)
ci

(0.15357196267702494, 0.047143907775453915)

In [43]:
ci[1] <= Dx0 <= ci[0]

True